server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java

```diff
@@ -85,27 +77,20 @@ public SetMessagesResponse process(SetMessagesRequest request,  MailboxSession m
     private void update(MessageId messageId, UpdateMessagePatch updateMessagePatch, MailboxSession mailboxSession,
                         SetMessagesResponse.Builder builder) {
         try {
-            MessageMapper messageMapper = mailboxSessionMapperFactory.createMessageMapper(mailboxSession);
-            Mailbox mailbox = mailboxMapperFactory.getMailboxMapper(mailboxSession)
-                    .findMailboxByPath(messageId.getMailboxPath());
-            Iterator<MailboxMessage> mailboxMessage = messageMapper.findInMailbox(
-                    mailbox, MessageRange.one(messageId.getUid()), MessageMapper.FetchType.Metadata, LIMIT_BY_ONE);
-            MailboxMessage messageWithUpdatedFlags = applyMessagePatch(messageId, mailboxMessage.next(), updateMessagePatch, builder);
-            savePatchedMessage(mailbox, messageId, messageWithUpdatedFlags, messageMapper);
+            MessageManager messageManager = mailboxManager.getMailbox(messageId.getMailboxPath(), mailboxSession);
+            MessageResultIterator message = messageManager.getMessages(messageId.getUidAsRange(), FetchGroupImpl.MINIMAL, mailboxSession);
+            updateFlags(messageId, updateMessagePatch, mailboxSession, messageManager, message.next());
+            builder.updated(ImmutableList.of(messageId));
         } catch (NoSuchElementException e) {
             addMessageIdNotFoundToResponse(messageId, builder);
         } catch (MailboxException e) {
             handleMessageUpdateException(messageId, builder, e);
         }
     }
 
-    private boolean savePatchedMessage(Mailbox mailbox, MessageId messageId,
-                                       MailboxMessage message,
-                                       MessageMapper messageMapper) throws MailboxException {
-        return messageMapper.updateFlags(mailbox, new FlagsUpdateCalculator(message.createFlags(),
-                        MessageManager.FlagsUpdateMode.REPLACE),
-                MessageRange.one(messageId.getUid()))
-                .hasNext();
+    private void updateFlags(MessageId messageId, UpdateMessagePatch updateMessagePatch, MailboxSession mailboxSession, MessageManager messageManager, MessageResult messageResult) throws MailboxException {
+        Flags newState = updateMessagePatch.applyToState(messageResult.getFlags());
```

Comment:
I guess I was not here to comment on this, but here the implementation is vulnerable to race conditions.

If user A supply two concurent requests : 
 - one updating the **Seen** flag
 - one updating the **Flagged** flag (marked as starred in thunderbird for instance)
As a replace mode is used here, the two request will not commute and the message will either be : 
 - Marked as unread but flagged
 - Marked as read but unflagged
Depending of the execution order....

**Why do we care?**

Because one usually make several updates of single flags of the same message in some very short time window (read + deleted // read + flagged // with longer window : read + answered). So I expect this problem to be seen.

Because it is avoidable... **UpdateMessagePatch** is actually a week reresentation that do not capture the fact that in JMAP these "flag" properties are actually independent from one an other.

**How to solve this?** (In my opinion...)

I agree Mailbox API do not fit the requirements right know to solve this problem, as Mailbox API do not provide other way to set and unset flags in a single request (unless you do several calls). But this could be corrected by making **FlagsUpdateCalculator** a higher level object, write for it a real builder (that catches what kind of flags updates one truely want to do with JMAP), and pass it by the **MessageManager** API.